Integrate OpenJCEPlus into Semeru OpenJDK

taoliult opened the pull request:
Depends on https://github.com/eclipse-openj9/openj9/pull/18544 being merged first.

Integrate the building of OpenJCEPlus with Semeru OpenJDK as a module "openjceplus". The java codes will be built with Semeru OpenJDK as a jmod, and the native codes will be built as a ".so" or ".dll" library.

The "bash get_source.sh" will download the OpenJCEPlus repo and also the ICC binaries which needed when building the OpenJCEPlus native codes.

Example:

bash get_source.sh -openjceplus-repo=[OpenJCEPlus repo] -openjceplus-branch=[OpenJCEPlus branch] -gskit-bin=[ICC binary tar file] -gskit-sdk-bin=[ICC SDK binary tar file] -gskit-credential=[Credential for downloading ICC]

bash configure --with-boot-jdk=[Boot JDK] --enable-openjceplus

Signed-off-by: Tao Liu [tao.liu@ibm.com](mailto:tao.liu@ibm.com)

taoliult commented:
@pshipton 

It is a draft PR, please help to review, if the makefiles which we updated are correct.

And in the “closed/custom/Main-pre.gmk”, I put the native codes building “openjceplus-libs” under the “java.base-copy”, same as what the zOS team did when they integrated the OpenJCEPlus building with Semeru. But, since we add a new module “openjceplus”, so I am not sure if under the “java.base-copy” is still good, or do we need another, like “openjceplus-copy” to tigger the OpenJCEPlus native codes building. Please help to review and advise.

taoliult commented:
@jasonkatonica @KostasTsiounis @WilburZjh @paritkum
FYI.

keithc-ca commented on closed/Components.gmk:
```diff
@@ -0,0 +1,71 @@
+# ===========================================================================
+# (c) Copyright IBM Corp. 2020, 2022 All Rights Reserved
```
Please ensure correct copyright dates are correct in all modified files (the second year should be 2023).

keithc-ca commented on closed/Components.gmk:
```diff
@@ -0,0 +1,71 @@
+	@$(ECHO) Staging ICC in $(OPENJCEPLUS_TOPDIR)/icc
+	$(call download_icc_binaries,$(GSKIT_BIN),$(GSKIT_SDK_BIN),$(GSKIT_BIN_CREDENTIAL))
+
+endif # BUILD_JGSKIT
```
Each text file should have a final newline.

keithc-ca commented on closed/autoconf/custom-hook.m4:
```diff
@@ -790,3 +791,70 @@ AC_DEFUN([OPENJ9_GENERATE_TOOL_WRAPPERS],
+  AC_ARG_WITH(gskit-bin-credential, [AS_HELP_STRING([--with-gskit-bin-credential], [specify path to the gskit bin credential])])
+
+  AC_MSG_CHECKING([if gskit and gskit sdk bin url is set])
+  if test "x$with_gskit_bin" = xno || test "x$with_gskit_sdk_bin" = xno ; then
+    AC_MSG_RESULT([no])
+  elif test "x$with_gskit_bin" = x || test "x$with_gskit_bin" = xyes || test "x$with_gskit_sdk_bin" = x || test "x$with_gskit_sdk_bin" = xyes ; then
```
Normally "" and "no" are given the same treatment. Other values (perhaps including "yes") should be taken as is.

keithc-ca commented on closed/autoconf/custom-hook.m4:
```diff
@@ -790,3 +791,70 @@ AC_DEFUN([OPENJ9_GENERATE_TOOL_WRAPPERS],
+  printf "with '--with-gskit-bin-credential' to specify the credential(if needed)\n"
+  printf "when downloading the GSKIT binary and GSKIT SDK binary.\n"
+  printf "\n"
+  printf "To ignore OpenJCEPLUs with GSKIT, run configure with '--with_gskit_bin=no',\n"
+  printf "or '--with-gskit-sdk-bin=no'."
```
The default should be to ignore OpenJCEPlus (typo in `OpenJCEPLUs`); those options should only be required to choose some non-default configuration.

keithc-ca commented on closed/autoconf/custom-spec.gmk.in:
```diff
@@ -181,3 +181,11 @@ J9JCL_SOURCES_DONEFILE := $(MAKESUPPORT_OUTPUTDIR)/j9jcl.done
 JVM_FEATURES_server :=
+
+# required by OpenJCEPlus
+BUILD_JGSKIT             := @BUILD_JGSKIT@
+GSKIT_BIN                := @GSKIT_BIN@
+GSKIT_SDK_BIN            := @GSKIT_SDK_BIN@
+GSKIT_BIN_CREDENTIAL     := @GSKIT_BIN_CREDENTIAL@
+GSKIT_HOME               := @GSKIT_HOME@
+OPENJCEPLUS_TOPDIR       := @OPENJCEPLUS_TOPDIR@
```
Please sort names in this group.

keithc-ca commented on closed/custom/Main-pre.gmk:
```diff
@@ -0,0 +1,42 @@
+openjceplus-libs :
+	@$(ECHO) Compiling OpenJCEPlus native
+	export PLATFORM=x86-linux64 JAVA_HOME=$(BOOT_JDK) GSKIT_HOME=$(GSKIT_HOME) JCE_CLASSPATH=$(OPENJCEPLUS_TOPDIR)/src/main/java/:$(JDK_OUTPUTDIR)/modules/java.base/ \
+		&& $(MAKE) -j1 -C $(OPENJCEPLUS_TOPDIR)/src/main/native/ -f jgskit.mak all
```
Why is `-j1` required here?
Why does it seem to assume `x86-linux64`; are other platforms not supported?

keithc-ca commented on closed/get_j9_source.sh:
```diff
@@ -39,6 +39,10 @@ usage() {
 	echo "  -omr-branch       the OpenJ9/omr git branch: openj9"
 	echo "  -omr-sha          a commit SHA for the omr repository"
 	echo "  -omr-reference    a local repo to use as a clone reference"
+	echo "  -openjceplus-repo the OpenJCEPlus repository url: git@github.ibm.com:JavaSecurity/OpenJCEPlus-with-GSKIT.git"
```
I suggest there should be no default for these options, but they certainly should not be IBM internal.

keithc-ca commented on closed/get_j9_source.sh:
```diff
@@ -62,6 +66,9 @@ branches[openj9]=master
 git_urls[omr]=https://github.com/eclipse-openj9/openj9-omr
 branches[omr]=openj9
 
+#git_urls[OpenJCEPlus]=https://github.com/taoliult/OpenJCEPlus
+#branches[OpenJCEPlus]=main
```
The defaults for these should be empty.

keithc-ca commented on get_source.sh:
```diff
@@ -40,6 +40,10 @@ usage() {
 	echo "  -omr-branch       the OpenJ9/omr git branch: openj9"
 	echo "  -omr-sha          a commit SHA for the omr repository"
 	echo "  -omr-reference    a local repo to use as a clone reference"
+	echo "  -openjceplus-repo the OpenJCEPlus repository url: git@github.ibm.com:JavaSecurity/OpenJCEPlus-with-GSKIT.git"
```
Again, no IBM internal references.

keithc-ca commented on make/common/JavaCompilation.gmk:
```diff
@@ -264,7 +264,7 @@ define SetupJavaCompilationBody
 
   # Allow overriding on the command line
-  JAVA_WARNINGS_ARE_ERRORS ?= -Werror
+  # JAVA_WARNINGS_ARE_ERRORS ?= -Werror
```
This is not acceptable: we don't want to ignore warnings. If the warnings are in OpenJCEPlus code, those should be addressed, or failing that, `openjceplus` could be added to the list of `WARNING_MODULES` in `custom-spec.gmk.in`.

keithc-ca commented on src/java.base/share/classes/module-info.java:
```diff
@@ -210,6 +211,7 @@
         jdk.charsets,
         jdk.compiler,
         jdk.crypto.cryptoki,
+        openjceplus,
```
Please make additions so the lists remain in order.

keithc-ca reviewed (changes requested):
How are these to be tested externally? Is there a github repository that could be used?

keithc-ca commented:
While working with Java 17 might provide a more stable development environment, changes for the head stream, and jdk21 will be needed before this can be merged here.

keithc-ca commented on closed/Components.gmk:
```diff
@@ -0,0 +1,71 @@
+# Download ICC binaries into $(OPENJCEPLUS_TOPDIR)/icc folder
+stage-openjceplus :
+	@$(ECHO) Staging OpenJCEPlus in $(TOPDIR)/src/openjceplus
+	$(call copy_component_tree,$(TOPDIR)/src/openjceplus/share/classes,$(OPENJCEPLUS_TOPDIR)/src/main/java/*)
```
This is not reasonable: it will pollute the working directory.

taoliult replied:
@keithc-ca I appended the OPENJCEPLUS_TOPDIR folder to TOP_SRC_DIRS in Modules.gmk, as below. I can see its java source codes and native codes built and in its "target" folder. But we would like the OpenJCEPlus java code built as a jmod file in JDK's "jmod" folder. Do you have any idea how we can do that?

```
TOP_SRC_DIRS += \
	$(J9JCL_SOURCES_DIR) \
	$(OPENJCEPLUS_TOPDIR) \
	#
```

keithc-ca replied:
The source in `$(OPENJCEPLUS_TOPDIR)` must be organized as expected by openjdk, with the same structure you'll find under `$(TOPDIR)/src`:
```
openjceplus/share/classes/module-info.java
openjceplus/share/classes/...         (other java code, etc.)
openjceplus/share/native/...          (if needed)
```

taoliult replied:
Copied the java codes into "$TOPDIR/OpenJCEPlus/src/main/openjceplus/share/classes/", and appended "$TOPDIR/OpenJCEPlus/src/main/" folder to "TOP_SRC_DIRS". The java codes built success and jmod file generated.

keithc-ca commented:
Please address the copyright check failures.

taoliult commented:
@keithc-ca 
For your question "How are these to be tested externally? Is there a github repository that could be used?". 
Right now, we are still working on the public repo setup. So, for now, we just use our internal or private repo for the test purpose.

taoliult commented on closed/custom/Main.gmk:
```diff
@@ -119,3 +119,6 @@ ifneq (,$(HEALTHCENTER_JAR))
 endif # HEALTHCENTER_JAR
+
+# include Main-pre.gmk
+$(eval $(call IncludeCustomExtension, Main-pre.gmk))
```
Moved the content of Main-pre.gmk into Main.gmk, and deleted Main-pre.gmk.

taoliult commented on closed/autoconf/custom-hook.m4:
```diff
@@ -790,3 +791,68 @@ AC_DEFUN([OPENJ9_GENERATE_TOOL_WRAPPERS],
+  printf "\n"
+  printf "Run configure with '--with-gskit-bin' and '--with-gskit-sdk-bin' to\n"
+  printf "specify the URL of GSKIT binary and GSKIT SDK binary. And run configure\n"
+  printf "with '--with-gskit-bin-credential' to specify the credential(if needed)\n"
```
All required source (including binaries) fetched by get_source.sh.

taoliult commented on closed/get_j9_source.sh:
```diff
@@ -39,6 +39,10 @@ usage() {
 	echo "  -omr-branch       the OpenJ9/omr git branch: openj9"
 	echo "  -omr-sha          a commit SHA for the omr repository"
 	echo "  -omr-reference    a local repo to use as a clone reference"
+	echo "  -openjceplus-repo the OpenJCEPlus repository url"
+	echo "  -openjceplus-branch the OpenJCEPlus git branch"
+	echo "  -openjceplus-sha  a commit SHA for the OpenJCEPlus repository"
+	echo "  -openjceplus-reference a local repo to use as a clone reference"
```
Added the options to specify necessary credentials and download gskit.

taoliult commented on closed/Components.gmk:
```diff
@@ -0,0 +1,71 @@
+        @echo "GSKIT_BIN_CREDENTIAL is either undefined or empty, will add curl command HERE without credential.")
+	@$(TAR) -xf $(OPENJCEPLUS_TOPDIR)/icc/jgsk_crypto_sdk.tar -C $(OPENJCEPLUS_TOPDIR)/icc/
+	@$(TAR) -xf $(OPENJCEPLUS_TOPDIR)/icc/jgsk_crypto.tar -C $(OPENJCEPLUS_TOPDIR)/icc/jgsk_sdk/lib64/
+endef
```
Moved the gskit download to get_j9_source.sh

taoliult commented on closed/autoconf/custom-hook.m4:
```diff
@@ -790,3 +791,68 @@ AC_DEFUN([OPENJ9_GENERATE_TOOL_WRAPPERS],
+  AC_ARG_WITH(gskit-bin-credential, [AS_HELP_STRING([--with-gskit-bin-credential], [specify path to the gskit bin credential])])
+
+  AC_MSG_CHECKING([if gskit and gskit sdk bin url is set])
+  if test "x$with_gskit_bin" = x || test "x$with_gskit_bin" = xyes || test "x$with_gskit_sdk_bin" = x || test "x$with_gskit_sdk_bin" = xyes ; then
```
No need to check gskit binaries here since moved the gskit download to get_j9_source.sh.

taoliult commented on closed/custom/common/Modules.gmk:
```diff
@@ -60,6 +60,7 @@ TOP_SRC_DIRS += \
 
 generate-j9jcl-sources $(J9JCL_SOURCES_DONEFILE) :
 	@+$(MAKE) $(MAKE_ARGS) -f $(TOPDIR)/closed/GensrcJ9JCL.gmk
+	@+$(COMPONENTS_MAKE) stage-components
```
Removed the "staging" part.

taoliult commented:
@keithc-ca For the code reviews, some of them updated the codes, some of them replied the questions. Please help to review and advise.

taoliult commented:
@keithc-ca Please help to review and advise. Thanks.

taoliult commented:
@keithc-ca 
I just updated this PR, by updating the Main.gmk for the OpenJCEPlus multi-platforms support. Right now, OpenJCEPlus supports 4 platforms, ppc64le_linux, ppc64_aix, x86-64_linux and x86-64_windows. Please help to review and advise. Thanks.

taoliult commented:
@keithc-ca This PR is ready for review, please help to review and advise.

keithc-ca commented on src/java.base/share/classes/module-info.java:
```diff
@@ -280,7 +283,8 @@
         java.net.http,
-        jdk.jartool;
+        jdk.jartool,
+        openjceplus;
```
This one is somewhat surprising; we really ought to limit the need for these qualified exports. Can you please double-check that _all_ the additions here are absolutely necessary?

taoliult replied:
I tried to remove this `sun.net.www `exports to openjceplus, but got below error. So, the `sun.net.www` package is needed when building openjceplus. 

```
/home/tao/mkJDK/openj9-openjdk-jdk17/OpenJCEPlus/src/main/openjceplus/share/classes/com/ibm/security/util/PropertyExpander.java:99: error: package sun.net.www is not visible
						val = sun.net.www.ParseUtil.encodePath(val);
						             ^
  (package sun.net.www is declared in module java.base, which does not export it to module openjceplus)
/home/tao/mkJDK/openj9-openjdk-jdk17/OpenJCEPlus/src/main/openjceplus/share/classes/com/ibm/security/util/PropertyExpander.java:102: error: package sun.net.www is not visible
					val = sun.net.www.ParseUtil.encodePath(val);
					             ^
  (package sun.net.www is declared in module java.base, which does not export it to module openjceplus)
```

But I re-checked all the exports in module-info, and removed the unnecessary exports for package `com.sun.crypto.provider`, `sun.security.jca`, `sun.security.provider`, `sun.security.x509` and `openj9.internal.security`.

keithc-ca commented on get_source.sh:
```diff
@@ -40,6 +40,13 @@ usage() {
+	echo "  -openjceplus-branch the OpenJCEPlus git branch"
+	echo "  -openjceplus-sha  a commit SHA for the OpenJCEPlus repository"
+	echo "  -openjceplus-reference a local repo to use as a clone reference"
+	echo "  -with-gskit-bin   the GSKit binary url"
+	echo "  -with-gskit-sdk-bin the GSKIT SDK binary url"
+	echo "  -with-gskit-bin-credential the credential for downloading the GSKit binaries"
```
Suggest the `-with` prefix is not needed for these options.
```
	echo "  -gskit-bin        the GSKit binary url"
	echo "  -gskit-sdk-bin    the GSKIT SDK binary url"
	echo "  -gskit-credential the credential for downloading the GSKit and GSKit SDK"
```
This also reflects the fact that the credentials are used for downloading both GSKit and SDK binaries.

keithc-ca commented on get_source.sh:
```diff
@@ -57,7 +64,7 @@ for i in "$@" ; do
 			;;
 
-		-openj9-repo=* | -openj9-branch=* | -openj9-sha=* | -openj9-reference=* | -omr-repo=* | -omr-branch=* | -omr-sha=* | -omr-reference=* | -parallel=* )
+		-openj9-repo=* | -openj9-branch=* | -openj9-sha=* | -openj9-reference=* | -omr-repo=* | -omr-branch=* | -omr-sha=* | -omr-reference=* | -openjceplus-repo=* | -openjceplus-branch=* | -openjceplus-sha=* | -openjceplus-reference=* | -with-gskit-bin=* | -with-gskit-sdk-bin=* | -with-gskit-bin-credential=* | -parallel=* )
```
The length of this line is getting out of hand, suggest it be folded (and sorted) like this:
```sh
		  -gskit-bin=* \
		| -gskit-credential=* \
		| -gskit-sdk-bin=* \
		| -omr-branch=* \
		| -omr-reference=* \
		| -omr-repo=* \
		| -omr-sha=* \
		| -openj9-branch=* \
		| -openj9-reference=* \
		| -openj9-repo=* \
		| -openj9-sha=* \
		| -openjceplus-branch=* \
		| -openjceplus-reference=* \
		| -openjceplus-repo=* \
		| -openjceplus-sha=* \
		| -parallel=* \
		)
```

keithc-ca commented on closed/custom/Main.gmk:
```diff
@@ -119,3 +119,53 @@ ifneq (,$(HEALTHCENTER_JAR))
+
+clean-openjceplus :
+	@$(ECHO) Cleaning OpenJCEPlus native
+	@$(RM) -f $(addprefix $(JGSKIT_MAKE_PATH), com_ibm_crypto_plus_provider_icc_NativeInterface.h com_ibm_crypto_plus_provider_icc_FastJNIBuffer.h libjgskit.x)
```
Please split this and sort the names:
```
	@$(RM) -f \
		$(addprefix $(JGSKIT_MAKE_PATH)/, \
			com_ibm_crypto_plus_provider_icc_FastJNIBuffer.h \
			com_ibm_crypto_plus_provider_icc_NativeInterface.h \
			libjgskit.x)
```
Note that `JGSKIT_MAKE_PATH` does not end with a `/`.

taoliult replied:
When building the OpenJCEPlus with Semeru OpenJDK, the generated class and native files are in the `/OpenJCEPlus/target/` directory. So, do you think we still need these clean codes? I removed those OpenJCEPlus clean codes for now.

taoliult commented on closed/custom/Main.gmk:
```diff
@@ -119,3 +119,53 @@ ifneq (,$(HEALTHCENTER_JAR))
+JCE_CLASSPATH := $(OPENJCEPLUS_TOPDIR)/src/main/java:$(TOPDIR)/src/java.base/share/classes
+JCE_CLASSPATH_WIN64 := "$(call MixedPath,$(OPENJCEPLUS_TOPDIR)/src/main/java)\;$(call MixedPath,$(TOPDIR)/src/java.base/share/classes)"
+
+ifeq ($(OPENJDK_TARGET_OS), aix)
```
Updated the codes according code review and moved the codes into closed/make/modules/openjceplus/Lib.gmk.

taoliult commented on closed/custom/Main.gmk:
```diff
@@ -119,3 +119,53 @@ ifneq (,$(HEALTHCENTER_JAR))
+JGSKIT_MAKE := jgskit.mak
+JGSKIT_MAKE_PATH := $(OPENJCEPLUS_TOPDIR)/src/main/native/
+JCE_CLASSPATH := $(OPENJCEPLUS_TOPDIR)/src/main/java:$(TOPDIR)/src/java.base/share/classes
+JCE_CLASSPATH_WIN64 := "$(call MixedPath,$(OPENJCEPLUS_TOPDIR)/src/main/java)\;$(call MixedPath,$(TOPDIR)/src/java.base/share/classes)"
```
Redefined the `JCE_CLASSPATH` but can not remove the escape for `\;`. Once remove the escape, when building on Windows, will get the following error:

```
[2023-11-21T04:50:50.465Z] error: no source files
[2023-11-21T04:50:50.465Z] /bin/sh: line 1: C:/users/jenkins/workspace/build_jdk17_x86-64_windows_personal/src/java.base/share/classes: Is a directory
[2023-11-21T04:50:50.465Z] make[4]: *** [jgskit.win64.mak:62: javah] Error 126
[2023-11-21T04:50:50.465Z] make[3]: *** [Lib.gmk:42: compile-libs] Error 2
[2023-11-21T04:50:50.465Z] make[2]: *** [make/Main.gmk:203: openjceplus-libs] Error 2
[2023-11-21T04:50:50.465Z] make[2]: *** Waiting for unfinished jobs....
```

taoliult commented on closed/custom/Main.gmk:
```diff
@@ -119,3 +119,53 @@ ifneq (,$(HEALTHCENTER_JAR))
+	@$(ECHO) Compiling OpenJCEPlus native
+	export PLATFORM=$(JGSKIT_TARGET) JAVA_HOME=$(BOOT_JDK) GSKIT_HOME=$(GSKIT_HOME) JCE_CLASSPATH=$(JCE_CLASSPATH) \
+		&& $(MAKE) -j1 -C $(JGSKIT_MAKE_PATH) -f $(JGSKIT_MAKE) all
+	@$(ECHO) OpenJCEplus compile complete
```
Moved the native code building into `closed/make/modules/openjceplus/Lib.gmk`

taoliult commented on closed/get_j9_source.sh:
```diff
@@ -173,6 +208,38 @@ if [ ${pflag} = true ] ; then
+	tar -xf ./icc/jgsk_crypto_sdk.tar -C ./icc/
+	tar -xf ./icc/jgsk_crypto.tar -C ./icc/jgsk_sdk/lib64/
+
+	# Create OpenJCEPlus Java module folder
+	mkdir -p ./src/main/openjceplus/share/classes/
+	cp -r ./src/main/java/* ./src/main/openjceplus/share/classes/
```
OpenJCEPlus also supports the Maven build and it follows the existing folder structure when building. If we change the current OpenJCEPlus java code folder structure to the structure same as the OpenJDK java module, there may many places in the OpenJCEPlus need to change. Consider our current OpenJCEPlus phase 1 timeline, I would suggest we just keep what it is, and in our next OpenJCEPlus phase, the phase 2 or phase 3, we can discuss and work on the OpenJCEPlus java code folder structure changes.

keithc-ca replied:
This isn't going to be pleasant for anyone actively working on OpenJCEPlus and the integration with openjdk. One should be able to modify any source files within OpenJCEPlus without manually managing the copies this creates. This makes it a more serious issue from my perspective.

taoliult replied:
@keithc-ca 
Yes, I understand. But for the current phase, due to the tight schedule, I would say we keep this source file structure, and we can modify this source file structure in our next phase. 

@jasonkatonica  any advice?

jasonkatonica replied:
As you mention above @taoliult , the OpenJCEPlus project is being built as a stand alone project that will allow it to be built and used even without bundling into a JDKs at all ( as a modular jar file ). Given that its a maven project the maven conventions are [currently documented](https://maven.apache.org/guides/introduction/introduction-to-the-standard-directory-layout.html) to have this type of structure. I dont really see the need to change the directory structure given its used in a bundled and non bundled form. We also do not anticipate ever needing to change that directory structure for OpenJCEPlus from `/src/main/java` to anything else so the maintenance costs should be very minimal on this line of code.

jasonkatonica replied:
Yes that is true, you are able to use this without bundling OpenJCEPlus with the JDK. This is one of our goals in OpenJCEPlus, to attract users outside of Semeru project itself as the project is a stand alone open source project. 

This being said users would have a hard time setting up their exact exports statements, configuring the modular jar correctly in the java.security file, auditing and enforcing FIPS related profiles, etc. 

The code we are putting in place within Semeru is to aid users in the use of OpenJCEPlusFIPS and in the future OpenJCEPlus ( non fips operations ) in a seamless as possible way.

IBM has a long history of bundling FIPS, JCE providers, and other crypto components within the runtime itself. Supporting products and users in this way is easier then supporting products and configurations using a stand alone provider.

keithc-ca replied:
I don't see the folder structure change I suggested as terribly onerous, I'm not sure why you resist that.

If OpenJCEPlus *can* be built without changes to OpenJDK, then I suggest we use that build flow. The _integration_ would simply be to "import" (I think that's the right OpenJDK concept) the module into the JDK. OpenJCEPlus would provide the build pieces to produce a `.jmod` file (or equivalent) that can be included among other `.jmod` files and consumed by `jlink` to produce the final image. See the `ibm.healthcenter` integration for a (loose) example of how that would work.

taoliult commented:
@keithc-ca 
Codes updated according to code review, and also answered some questions. Right now, the codes are ready for review. Please help to review and advise.

pshipton commented:
jenkins compile amac jdk17

pshipton commented:
Atm this doesn't build without openjceplus so it can't be merged as-is until we have openjceplus available. Will openjceplus be supported and included on all platforms?

jasonkatonica commented:
> Atm this doesn't build without openjceplus so it can't be merged as-is until we have openjceplus available. Will openjceplus be supported and included on all platforms?

Hi Peter, Unfortunately OpenJCEPlusFips wont be available on all platforms for some time. We will have to look into a workaround as to how to allow the build to pass on platforms where OpenJCEPlusFips wont be available in the near term.

taoliult commented:
> Atm this doesn't build without openjceplus so it can't be merged as-is until we have openjceplus available. Will openjceplus be supported and included on all platforms?

@pshipton 
Right now, the OpenJCEPlus on Semeru only supports x86-64_linux, ppc64le_linux, x86-64_windows and ppc64_aix platforms. So, about the export to openjceplus in module-info.java file, can we do the similar as what CRIU did (below), adding the pre-processor checks in “openj9/jcl/src/java.base/share/classes/module-info.java.extra”, to only export to openjceplus in module-info.java file when building on the supported target platforms?

```
/*[IF CRIU_SUPPORT]*/
exports openj9.internal.criu to
    openj9.criu;
exports openj9.internal.criu.security to
    openj9.criu;
exports jdk.internal.misc to
    openj9.criu;
opens jdk.internal.misc to
    openj9.criu;
opens java.lang to
    openj9.criu;
/*[ENDIF] CRIU_SUPPORT */
```

If we can do the similar, do you have any document or example about how to do/define it?

pshipton commented:
See
https://github.com/eclipse-openj9/openj9/blob/master/jcl/jpp_configuration.xml#L40
https://github.com/ibmruntimes/openj9-openjdk-jdk17/blob/openj9/closed/JPP.gmk#L29-L31
https://github.com/ibmruntimes/openj9-openjdk-jdk17/blob/openj9/closed/OpenJ9.gmk#L295-L300
https://github.com/ibmruntimes/openj9-openjdk-jdk17/blob/openj9/closed/OpenJ9.gmk#L463-L467
https://github.com/ibmruntimes/openj9-openjdk-jdk17/blob/openj9/closed/autoconf/custom-hook.m4#L44
https://github.com/ibmruntimes/openj9-openjdk-jdk17/blob/openj9/closed/autoconf/custom-hook.m4#L371-L396
https://github.com/ibmruntimes/openj9-openjdk-jdk17/blob/openj9/closed/autoconf/custom-spec.gmk.in#L72
https://github.com/ibmruntimes/openj9-openjdk-jdk17/blob/openj9/closed/custom/Images-pre.gmk#L22
https://github.com/ibmruntimes/openj9-openjdk-jdk17/blob/openj9/closed/custom/Main.gmk#L57

taoliult commented:
Added the “OPENJCEPLUS_SUPPORT” into JPP_TAGS in file “closed/JPP.gmk”. And also removed all the exports to openjceplus module in “src/java.base/share/classes/module-info.java”

Run the builds on the OpenJCEPlus not supported platforms, s390x_linux and aarch64_linux, links below. The builds look good, no openjceplus module generated and no exports to unknown module error.

https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/19529/
https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/19531/

But before this PR merged, need to review and merge the OpenJ9 PR https://github.com/eclipse-openj9/openj9/pull/18544 first. 

@pshipton Please help to review and advise.

pshipton commented:
Pls add the openjceplus module to the JRE build.
https://github.com/ibmruntimes/openj9-openjdk-jdk17/blob/openj9/closed/custom/Images-pre.gmk#L22

keithc-ca commented on closed/JPP.gmk:
```diff
@@ -30,6 +30,10 @@ ifeq (true,$(OPENJ9_ENABLE_CRIU_SUPPORT))
   JPP_TAGS += CRIU_SUPPORT
 endif # OPENJ9_ENABLE_CRIU_SUPPORT
 
+ifeq (true,$(BUILD_OPENJCEPLUS))
+  JPP_TAGS += OPENJCEPLUS_SUPPORT
+endif # BUILD_OPENJCEPLUS
```
Please move this after `INLINE-TYPES` so the names in `JPP_TAGS` remain sorted.

keithc-ca commented on closed/custom/Main.gmk:
```diff
@@ -119,3 +119,5 @@ ifneq (,$(HEALTHCENTER_JAR))
   ibm.healthcenter-java : ibm.healthcenter-copy
 endif # HEALTHCENTER_JAR
+
+openjceplus-copy : openjceplus-libs
```
I think more is needed here so that `make clean-openjceplus` will do the right thing.

taoliult replied:
@keithc-ca 

The clean codes, as below, will use the variable, for example "$(JGSKIT_MAKE)",  which defined in Lib.gmk. Do you have any suggestion, for example, how can we add the below clean codes in Lib.gmk, or any better way to get this variable "$(JGSKIT_MAKE)" in the Main.gmk?

```
.PHONY : clean-openjceplus

clean : clean-openjceplus

clean-openjceplus :
	@$(ECHO) Cleaning OpenJCEPlus native
	@$(RM) -f \
		$(addprefix $(JGSKIT_MAKE_PATH)/, \
			com_ibm_crypto_plus_provider_icc_FastJNIBuffer.h \
			com_ibm_crypto_plus_provider_icc_NativeInterface.h \
			libjgskit.*)
	$(MAKE) -C $(JGSKIT_MAKE_PATH) -f $(JGSKIT_MAKE) cleanAll
```

keithc-ca commented on closed/make/modules/openjceplus/Copy.gmk:
```diff
@@ -0,0 +1,49 @@
+include $(TOPDIR)/closed/CopySupport.gmk
+
+ifeq (true,$(BUILD_OPENJCEPLUS))
+  # Copy OpenJCEPlus native libraries
```
Please finish each sentence with a period.

keithc-ca commented on closed/make/modules/openjceplus/Copy.gmk:
```diff
@@ -0,0 +1,49 @@
+
+ifeq (true,$(BUILD_OPENJCEPLUS))
+  # Copy OpenJCEPlus native libraries
+  $(eval $(call SetupCopyFiles, COPY_JGSKIT_LIBS, \
+      SRC := $(OPENJCEPLUS_TOPDIR)/target, \
+      FILES := $(filter %.dll %.so %.x, $(call FindFiles, $(OPENJCEPLUS_TOPDIR)/target)), \
+      FLATTEN := true, \
+      DEST := $(LIB_DST_DIR), \
+  ))
+
+  TARGETS += $(COPY_JGSKIT_LIBS)
+
+  # Bundle GSKIT library
+  OCK_DIR := $(TOPDIR)/OpenJCEPlus/ock/jgsk_sdk/lib64
+  ifeq ($(call isTargetOs, windows), true)
+    OCK_SUB_DIR := modules_cmds
+  else
+    OCK_SUB_DIR := modules_libs
+  endif
+
```
I think the makefiles should restrict themselves to a single root name (suggest `OPENJCEPLUS`) rather than also use `GSKIT`, `JGSKIT`, and `OCK`.

keithc-ca commented on closed/make/modules/openjceplus/Lib.gmk:
```diff
@@ -0,0 +1,64 @@
+ifeq ($(call isTargetOs, aix), true)
+  JGSKIT_TARGET := ppc-aix64
+else ifeq ($(call isTargetOs, linux), true)
+  ifeq ($(call isTargetCpu, x86_64), true)
+    JGSKIT_TARGET := x86-linux64
+  else  ifeq ($(call isTargetCpu, ppc64le), true)
```
Please consider target CPUs in alphabetical order (ppc64le before x86_64).

keithc-ca commented on closed/make/modules/openjceplus/Lib.gmk:
```diff
@@ -0,0 +1,64 @@
+ifeq (true,$(BUILD_OPENJCEPLUS))
+
+# Identify the desired JGSKIT target platform.
+GSKIT_HOME := $(TOPDIR)/OpenJCEPlus/ock/jgsk_sdk
```
This should use `$(OPENJCEPLUS_TOPDIR)` rather than `$(TOPDIR)/OpenJCEPlus`.

keithc-ca commented on closed/make/modules/openjceplus/Lib.gmk:
```diff
@@ -0,0 +1,64 @@
+GSKIT_HOME := $(TOPDIR)/OpenJCEPlus/ock/jgsk_sdk
+JGSKIT_TARGET :=
+JGSKIT_MAKE := jgskit.mak
+JGSKIT_MAKE_PATH := $(OPENJCEPLUS_TOPDIR)/src/main/native/
```
Please remove the trailing slash.

keithc-ca commented on closed/make/modules/openjceplus/Lib.gmk:
```diff
@@ -0,0 +1,64 @@
+
+compile-libs :
+	@$(ECHO) Compiling OpenJCEPlus native
+	export PLATFORM=$(JGSKIT_TARGET) JAVA_HOME=$(BOOT_JDK) GSKIT_HOME=$(GSKIT_HOME) JCE_CLASSPATH=$(JCE_CLASSPATH) \
+		&& $(MAKE) -j1 -C $(JGSKIT_MAKE_PATH) -f $(JGSKIT_MAKE) all
```
Please split this and list the environment variable assignments in alphabetical order:
```
	export \
			GSKIT_HOME=$(GSKIT_HOME) \
			JAVA_HOME=$(BOOT_JDK) \
			JCE_CLASSPATH=$(JCE_CLASSPATH) \
			PLATFORM=$(JGSKIT_TARGET) \
		&& $(MAKE) -j1 -C $(JGSKIT_MAKE_PATH) -f $(JGSKIT_MAKE) all
```

keithc-ca commented on closed/make/modules/openjceplus/Lib.gmk:
```diff
@@ -0,0 +1,64 @@
+JGSKIT_TARGET :=
+JGSKIT_MAKE := jgskit.mak
+JGSKIT_MAKE_PATH := $(OPENJCEPLUS_TOPDIR)/src/main/native/
+JCE_CLASSPATH := $(OPENJCEPLUS_TOPDIR)/src/main/java:$(TOPDIR)/src/java.base/share/classes
```
This should refer to where the `java.base` class files are found rather than where (a subset of) the source files live.

taoliult replied:
@keithc-ca Do you know which path variable in the OpenJDK makefile is used for point to the `java.base` class files path?

keithc-ca replied:
I would start with `$(JDK_OUTPUTDIR)/modules/java.base` (e.g. see `DDR.gmk`), but it might eventually make sense to use `-source-path` and `-module-path`.

taoliult replied:
@keithc-ca 
This `JCE_CLASSPATH` is used when building the OpenJCEPlus native codes. I tried to change it to below.

`JCE_CLASSPATH := $(OPENJCEPLUS_TOPDIR)/target/classes:$(SUPPORT_OUTPUTDIR)/j9jcl/java.base/share/classes`

But get the following error. I think, when building the OpenJCEPlus native codes, the openjceplus java module and the java.base class files are not generated yet. And also, the native codes are not a part of openjceplus.jmod module, So, get this `package jdk.internal.logger is declared in module java.base, which does not export it to the unnamed module` error.

```
Compiling OpenJCEPlus native
/home/tao/mkJDK/openj9-openjdk-jdk17/OpenJCEPlus/src/main/java/com/ibm/misc/Debug.java:15: error: package jdk.internal.logger is not visible
import jdk.internal.logger.SimpleConsoleLogger; 
                   ^
  (package jdk.internal.logger is declared in module java.base, which does not export it to the unnamed module)
/home/tao/mkJDK/openj9-openjdk-jdk17/OpenJCEPlus/src/main/java/com/ibm/misc/Debug.java:16: error: package sun.util.logging is not visible
import sun.util.logging.PlatformLogger;
```

keithc-ca commented on src/java.base/share/classes/module-info.java:
```diff
@@ -182,6 +182,9 @@
         jdk.compiler,
         jdk.jlink;
     exports jdk.internal.logger to
+/*[IF OPENJCEPLUS_SUPPORT]*/
+        openjceplus,
+/*[ENDIF] OPENJCEPLUS_SUPPORT */
```
It should be a priority to eliminate the use of internal classes by OpenJCEPlus.

taoliult replied:
@keithc-ca 
Yes, we are working on the elimination of the use of the internal classes by OpenJCEPlus. But for the current phase, we still have to use those internal classes. 
@jasonkatonica FYI.

jasonkatonica replied:
The team faced a decision of maintaining 100k lines of code that was a duplicate of the sun internal classes or use some of the sun internal classes. A decision was made to use the sun internal classes. We are however in the long run planning on trying to find alternates and more and more external functions that could be used as an equivalent wherever possible. 

This particular package is something we certainly plan on removing as there are many alternatives to the jdk.internal.logger package. It will however not be done in this phase of development.

keithc-ca commented on closed/get_j9_source.sh:
```diff
@@ -39,6 +39,13 @@ usage() {
+	echo "  -openjceplus-branch the OpenJCEPlus git branch"
+	echo "  -openjceplus-sha  a commit SHA for the OpenJCEPlus repository"
+	echo "  -openjceplus-reference a local repo to use as a clone reference"
+	echo "  -with-gskit-bin   the GSKit binary url"
+	echo "  -with-gskit-sdk-bin the GSKIT SDK binary url"
+	echo "  -with-gskit-bin-credential the credential for downloading the GSKit binaries"
```
Please correct the help (the `-with` prefix has been removed).

keithc-ca commented on closed/make/modules/openjceplus/Lib.gmk:
```diff
@@ -0,0 +1,64 @@
+
+# Identify the desired JGSKIT target platform.
+GSKIT_HOME := $(TOPDIR)/OpenJCEPlus/ock/jgsk_sdk
+JGSKIT_TARGET :=
```
Given how this is used, `PLATFORM` would be a better suffix than `TARGET`.

taoliult commented:
@keithc-ca Codes updated according to code review, please help to review and advice. @jasonkatonica FYI.

taoliult commented:
Codes updated according to the code review, the x86 linux build looks good when building in my fyre VM. Committed the codes to run the builds on all the supported platforms. Once the builds are good, then I will update the comment for code review.

keithc-ca commented on closed/custom/Main.gmk:
```diff
@@ -119,3 +119,23 @@ ifneq (,$(HEALTHCENTER_JAR))
+.PHONY : openjceplus-clean
+
+openjceplus-clean :
+	@$(ECHO) Cleaning OpenJCEPlus native
```
Please fix this message (it needs a noun); suggest
```
	@$(ECHO) Cleaning OpenJCEPlus native code
```

keithc-ca commented on closed/get_j9_source.sh:
```diff
@@ -173,6 +208,38 @@ if [ ${pflag} = true ] ; then
+	mkdir -p ock/jgsk_sdk/lib64
+
+	if [ -n "$gskit_credential" ] ; then
+		# Download GSKit binaries from artifactory.
```
The reference to "artifactory" isn't necessarily appropriate; the URLs may point elsewhere.
```
		# Download GSKit binaries.
```
or just remove the comment.

keithc-ca commented on closed/get_j9_source.sh:
```diff
@@ -173,6 +208,38 @@ if [ ${pflag} = true ] ; then
+		exit 1
+	fi
+
+	tar -xf ./ock/jgsk_crypto_sdk.tar -C ./ock/
+	tar -xf ./ock/jgsk_crypto.tar -C ./ock/jgsk_sdk/lib64/
+
+	# Create OpenJCEPlus Java module folder.
+	mkdir -p ./src/main/openjceplus/share/classes/
+	cp -r ./src/main/java/* ./src/main/openjceplus/share/classes/
```
Please remove prefix `./` and (most) trailing slashes:
```
	tar -xf ock/jgsk_crypto_sdk.tar -C ock
	tar -xf ock/jgsk_crypto.tar -C ock/jgsk_sdk/lib64

	# Create OpenJCEPlus Java module folder.
	mkdir -p src/main/openjceplus/share/classes
	cp -r src/main/java/* src/main/openjceplus/share/classes/

keithc-ca commented on closed/make/modules/openjceplus/Copy.gmk:
```diff
@@ -0,0 +1,49 @@
+  TARGETS += $(OPENJCEPLUS_JGSKIT_LIBS_COPY)
+
+  # Bundle GSKIT library.
+  OPENJCEPLUS_OCK_DIR := $(TOPDIR)/OpenJCEPlus/ock/jgsk_sdk/lib64
```
Please use `OPENJCEPLUS_TOPDIR`:
```
  OPENJCEPLUS_OCK_DIR := $(OPENJCEPLUS_TOPDIR)/ock/jgsk_sdk/lib64
```

keithc-ca commented on closed/make/modules/openjceplus/Lib.gmk:
```diff
@@ -0,0 +1,68 @@
+ifeq (true,$(BUILD_OPENJCEPLUS))
+
+# Identify the desired JGSKIT target platform.
+GSKIT_HOME := $(OPENJCEPLUS_TOPDIR)/ock/jgsk_sdk
```
Please consistently use the prefix `OPENJCEPLUS_`:
```
OPENJCEPLUS_GSKIT_HOME := $(OPENJCEPLUS_TOPDIR)/ock/jgsk_sdk
```
Also please list definitions in alphabetical order.

keithc-ca commented on closed/make/modules/openjceplus/Lib.gmk:
```diff
@@ -0,0 +1,68 @@
+    OPENJCEPLUS_JGSKIT_MAKE := jgskit.win64.mak
+    OPENJCEPLUS_JCE_CLASSPATH := "$(call MixedPath,$(OPENJCEPLUS_TOPDIR)/src/main/java)\;$(call MixedPath,$(TOPDIR)/src/java.base/share/classes)"
+    GSKIT_HOME := $(call MixedPath,$(GSKIT_HOME))
+    BOOT_JDK := $(call MixedPath,$(BOOT_JDK))
```
It's not reasonable to redefine `BOOT_JDK` (from `spec.gmk`); a default definition for other platforms:
```
OPENJCEPLUS_BOOT_JDK := $(BOOT_JDK)
```
can be overridden for Windows
```
    OPENJCEPLUS_BOOT_JDK := $(call MixedPath,$(OPENJCEPLUS_BOOT_JDK))
```

keithc-ca commented on closed/make/modules/openjceplus/Lib.gmk:
```diff
@@ -0,0 +1,68 @@
+.PHONY : compile-libs
+
+compile-libs :
+	@$(ECHO) Compiling OpenJCEPlus native
```
Again, a noun is needed here
```
	@$(ECHO) Compiling OpenJCEPlus native code
```

taoliult commented:
Build looks good in my x86 linux fyre VM. Commit the changes to run the builds on all the supported platforms.

keithc-ca commented on closed/make/modules/openjceplus/Lib.gmk:
```diff
@@ -0,0 +1,69 @@
+  ifeq ($(call isTargetCpu, x86_64), true)
+    OPENJCEPLUS_BOOT_JDK := $(call MixedPath,$(OPENJCEPLUS_BOOT_JDK))
+    OPENJCEPLUS_GSKIT_HOME := $(call MixedPath,$(OPENJCEPLUS_GSKIT_HOME))
+    OPENJCEPLUS_JCE_CLASSPATH := "$(call MixedPath,$(OPENJCEPLUS_TOPDIR)/src/main/java)\;$(call MixedPath,$(TOPDIR)/src/java.base/share/classes)"
```
This is not consistent with other platforms; I think we want:
```
    OPENJCEPLUS_JCE_CLASSPATH := "$(call MixedPath,$(JDK_OUTPUTDIR)/modules/openjceplus\;$(JDK_OUTPUTDIR)/modules/java.base)"
```

taoliult replied:
yes, sorry about this, changed to make it consistent with above `OPENJCEPLUS_JCE_CLASSPATH`.

taoliult replied:
But, do we need to change to below, by calling `MixedPath` twice?

`OPENJCEPLUS_JCE_CLASSPATH := "$(call MixedPath,$(JDK_OUTPUTDIR)/modules/openjceplus)\;$(call MixedPath,$(JDK_OUTPUTDIR)/modules/java.base)"`

taoliult replied:
ok, updated to calling `MixedPath` twice.

keithc-ca commented:
After fixing the use of MixedPath, please squash and fix the commit message. Please only use ASCII quotes, (e.g. not “.so”) and put the "Signed-off-by:" line after a blank line at the end (not in the middle).

keithc-ca reviewed (approved):
This should be used as a model for pull requests to newer versions, e.g. https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/710. Changes for newer versions should be merged first.

I added @pshipton as a reviewer so this and related PRs can proceed during my absence.

taoliult commented:
@keithc-ca   Thanks so much for the review. I am working on the JDK next PR https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/710. Will update all the new changes from here to the JDK next PR.

taoliult commented:
@keithc-ca @pshipton 

Jenkins builds look good for all the supported platforms.

x86-64_linux:
https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/19799/

ppc64_aix:
https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/19800/

ppc64le_linux:
https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/19801/

x86-64_windows:
https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/19802/

Also run the build on OpenJCEPlus not supported platform s390x_linux, the build looks good:
https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/19811/

@jasonkatonica FYI.
